File: common/src/main/java/dev/mayaqq/estrogen/registry/advancements/triggers/KilledWithEffectTrigger.java
package dev.mayaqq.estrogen.registry.advancements.triggers;

import com.google.gson.JsonObject;
import dev.mayaqq.estrogen.Estrogen;
import net.minecraft.advancements.critereon.*;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.effect.MobEffect;
import net.minecraft.world.entity.Entity;

public class KilledWithEffectTrigger extends SimpleCriterionTrigger<KilledWithEffectTrigger.TriggerInstance> {
    static final ResourceLocation ID = Estrogen.id("killed_with_effect");

    @Override
    protected TriggerInstance createInstance(JsonObject json, ContextAwarePredicate predicate, DeserializationContext deserializationContext) {
        return new TriggerInstance(EntityPredicate.fromJson(json, "entity", deserializationContext), BuiltInRegistries.MOB_EFFECT.getHolder(json.get("mobeffect").getAsInt()).get().value(), predicate);
    }

    public void trigger(ServerPlayer player, Entity entityType) {
        this.trigger(player, (instance) -> instance.matches(player, entityType));
    }

    @Override
    public ResourceLocation getId() {
        return ID;
    }

    public static class TriggerInstance extends AbstractCriterionTriggerInstance {

        private final ContextAwarePredicate entity;
        private final MobEffect effect;

        public TriggerInstance(ContextAwarePredicate entity, MobEffect effect, ContextAwarePredicate player) {
            super(ID, player);
            this.entity = entity;
            this.effect = effect;
        }

        public boolean matches(ServerPlayer player, Entity entity) {
            if (this.entity.matches(EntityPredicate.createContext(player, entity))) {
                if (player.hasEffect(effect)) {
                    return true;
                }
            }
            return false;
        }

        public static KilledWithEffectTrigger.TriggerInstance killedWithEffect(ContextAwarePredicate entity, MobEffect effect, ContextAwarePredicate player) {
            return new KilledWithEffectTrigger.TriggerInstance(entity, effect, player);
        }

        @Override
        public JsonObject serializeToJson(SerializationContext context) {
            JsonObject jsonObject = super.serializeToJson(context);
            jsonObject.add("entity", this.entity.toJson(context));
            jsonObject.addProperty("mobeffect", BuiltInRegistries.MOB_EFFECT.getId(effect));
            return jsonObject;
        }
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/content/advancements/triggers/KilledWithEffectTrigger.kt
@file:EventSubscriber
package dev.mayaqq.estrogen.content.advancements.triggers

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.mayaqq.cynosure.events.api.EventSubscriber
import dev.mayaqq.cynosure.events.api.Subscription
import dev.mayaqq.cynosure.events.entity.LivingEntityEvent
import dev.mayaqq.estrogen.content.AdvancementTriggers
import net.minecraft.advancements.critereon.*
import net.minecraft.core.Holder
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.entity.Entity
import java.util.Optional

class KilledWithEffectTrigger : SimpleCriterionTrigger<KilledWithEffectTrigger.TriggerInstance>() {
    override fun codec(): Codec<TriggerInstance> = TriggerInstance.CODEC

    fun trigger(player: ServerPlayer, entityType: Entity) {
        this.trigger(
            player
        ) { instance: TriggerInstance ->
            instance.matches(
                player,
                entityType
            )
        }
    }

    @JvmRecord
    data class TriggerInstance(val entity: ContextAwarePredicate, val mobEffect: Holder<MobEffect>, val player: ContextAwarePredicate) : SimpleInstance {
        fun matches(player: ServerPlayer, entity: Entity): Boolean {
            if (this.entity.matches(EntityPredicate.createContext(player, entity))) {
                if (player.hasEffect(mobEffect)) {
                    return true
                }
            }
            return false
        }

        override fun player(): Optional<ContextAwarePredicate> = Optional.of(player)


        companion object {
            val CODEC: Codec<TriggerInstance> = RecordCodecBuilder.create { instance -> instance.group(
                    ContextAwarePredicate.CODEC.fieldOf("entity").forGetter(TriggerInstance::entity),
                    MobEffect.CODEC.fieldOf("mobEffect").forGetter(TriggerInstance::mobEffect),
                    EntityPredicate.ADVANCEMENT_CODEC.fieldOf("player").forGetter(TriggerInstance::player)
                ).apply(instance, ::TriggerInstance)
            }

            fun killedWithEffect(entity: ContextAwarePredicate, effect: Holder<MobEffect>, player: ContextAwarePredicate) : TriggerInstance {
                return TriggerInstance(entity, effect, player)
            }
        }
    }
}

@Subscription
fun onEntityDeath(event: LivingEntityEvent.Death) {
    if (event.source.entity is ServerPlayer) AdvancementTriggers.KilledWithEffect.trigger(event.source.entity as ServerPlayer, event.entity)
}